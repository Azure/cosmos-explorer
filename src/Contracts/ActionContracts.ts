/**
 * Data Explorer tab kinds
 */
export enum TabKind {
  SQLDocuments,
  MongoDocuments,
  MongoSchema,
  TableEntities,
  Graph,
  SQLQuery,
  ScaleSettings,
}

/**
 * Data Explorer pane kinds
 */
export enum PaneKind {
  AddCollection,
  CassandraAddCollection,
  DeleteCollection,
  DeleteDatabase,
  GlobalSettings,
  AdHocAccess,
  SwitchDirectory,
}

/**
 * Parameters to pass to DataExplorer in order to have it perform a given action
 */
export interface DataExplorerAction {
  actionType: ActionType | string;
}

/**
 * Open tab action
 */
export interface OpenTab extends DataExplorerAction {
  tabKind: TabKind | string;
}

/**
 * Open collection tab action
 */
export interface OpenCollectionTab extends OpenTab {
  databaseResourceId: string;
  collectionResourceId: string;
}

/**
 * Open query tab action
 */
export interface OpenQueryTab extends OpenCollectionTab {
  query: QueryInfo;
}

/**
 * Query info will be looked at in the order given in this interface (i.e. if text is provided, everything else is ignored)
 */
export interface QueryInfo {
  text?: string;
  partitionKeys?: string[];
}

/**
 * Open pane action
 */
export interface OpenPane extends DataExplorerAction {
  paneKind: PaneKind | string;
}

export interface OpenSampleNotebook extends DataExplorerAction {
  path: string;
}

/**
 * The types of actions that the DataExplorer supports performing upon opening.
 */
export enum ActionType {
  OpenTab,
  OpenCollectionTab,
  OpenPane,
  TransmitCachedData,
  OpenSampleNotebook,
}
