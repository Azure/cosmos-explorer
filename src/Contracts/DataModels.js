"use strict";
exports.__esModule = true;
exports.SparkClusterEndpointKind = exports.ConflictResolutionMode = exports.ApiKind = void 0;
var ApiKind;
(function (ApiKind) {
    ApiKind[ApiKind["SQL"] = 0] = "SQL";
    ApiKind[ApiKind["MongoDB"] = 1] = "MongoDB";
    ApiKind[ApiKind["Table"] = 2] = "Table";
    ApiKind[ApiKind["Cassandra"] = 3] = "Cassandra";
    ApiKind[ApiKind["Graph"] = 4] = "Graph";
    ApiKind[ApiKind["MongoDBCompute"] = 5] = "MongoDBCompute";
})(ApiKind = exports.ApiKind || (exports.ApiKind = {}));
var ConflictResolutionMode;
(function (ConflictResolutionMode) {
    ConflictResolutionMode["Custom"] = "Custom";
    ConflictResolutionMode["LastWriterWins"] = "LastWriterWins";
})(ConflictResolutionMode = exports.ConflictResolutionMode || (exports.ConflictResolutionMode = {}));
var SparkClusterEndpointKind;
(function (SparkClusterEndpointKind) {
    SparkClusterEndpointKind["SparkUI"] = "SparkUI";
    SparkClusterEndpointKind["HistoryServerUI"] = "HistoryServerUI";
    SparkClusterEndpointKind["Livy"] = "Livy";
})(SparkClusterEndpointKind = exports.SparkClusterEndpointKind || (exports.SparkClusterEndpointKind = {}));
