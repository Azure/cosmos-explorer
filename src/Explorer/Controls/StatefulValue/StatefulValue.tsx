/*
  Class used to track baseline and current values of a state object. If both the values differ,
  isDirty is set to true. The isDirty value is used to change css of the component and to govern logic
  around whether the current value should be persistently saved or not.
*/

export class StatefulValue<T> {
  public current: T;
  public baseline: T;
  public isValid: boolean;

  constructor(baselineValue?: T) {
    this.current = baselineValue;
    this.baseline = baselineValue;
    this.isValid = true;
  }

  public isDirty(): boolean {
    const currentStringValue = this.getStringValue(this.current);
    const baselineStringValue = this.getStringValue(this.baseline);

    return currentStringValue !== baselineStringValue;
  }

  private getStringValue = (value: T): string => {
    const type = typeof value;
    switch (type) {
      case "string":
      case "undefined":
      case "number":
      case "boolean":
        return value?.toString();

      default:
        return JSON.stringify(value);
    }
  };
}
