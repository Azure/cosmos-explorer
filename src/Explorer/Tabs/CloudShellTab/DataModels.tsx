/**
 * Copyright (c) Microsoft Corporation.  All rights reserved.
 */

export const enum OsType {
    Linux = "linux",
    Windows = "windows"
}

export const enum ShellType {
    Bash = "bash",
    PowerShellCore = "pwsh"
}
  
export const enum NetworkType {
    Default = "Default",
    Isolated = "Isolated"
}

export const enum SessionType {
    Mounted = "Mounted",
    Ephemeral = "Ephemeral"
}

export type Settings = {
    location: string;
    sessionType: SessionType;
    osType: OsType;
};

export type ProvisionConsoleResponse = {
    properties: {
        osType: OsType;
        provisioningState: string;
        uri: string;
    };
};

export type Authorization = {
    token: string;
};

export type ConnectTerminalResponse = {
    id: string;
    idleTimeout: string;
    rootDirectory: string;
    socketUri: string;
    tokenUpdated: boolean;
};


