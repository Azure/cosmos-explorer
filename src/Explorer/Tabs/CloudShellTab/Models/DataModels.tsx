export const enum OsType {
  Linux = "linux",
  Windows = "windows",
}

export const enum ShellType {
  Bash = "bash",
  PowerShellCore = "pwsh",
}

export const enum NetworkType {
  Default = "Default",
  Isolated = "Isolated",
}

/**
 * Azure CloudShell session types:
 * - Mounted: Sessions with persistent storage via an Azure File Share mount.
 *   Files and configurations are preserved between sessions, allowing for
 *   continuity of work across multiple CloudShell sessions.
 *
 * - Ephemeral: Temporary sessions without persistent storage.
 *   All files and changes are discarded when the session ends.
 *   These sessions start faster but don't retain user data.
 *
 * The session type affects resource allocation, startup time,
 * and whether user files/configurations persist between sessions.
 */
export const enum SessionType {
  Mounted = "Mounted",
  Ephemeral = "Ephemeral",
}

export type CloudShellSettings = {
  properties: UserSettingProperties;
};

export type UserSettingProperties = {
  networkType: string;
  preferredLocation: string;
  preferredOsType: OsType;
  preferredShellType: ShellType;
  userSubscription: string;
  sessionType: SessionType;
  vnetSettings: object;
};

export type ProvisionConsoleResponse = {
  properties: {
    osType: OsType;
    provisioningState: string;
    uri: string;
  };
};

export type Authorization = {
  token: string;
};

export type ConnectTerminalResponse = {
  id: string;
  idleTimeout: string;
  rootDirectory: string;
  socketUri: string;
  tokenUpdated: boolean;
};

export type ProviderAuthorization = {
  applicationId: string;
  roleDefinitionId: string;
};

export type ProviderResourceType = {
  resourceType: string;
  locations: string[];
  apiVersions: string[];
  defaultApiVersion?: string;
  capabilities?: string;
};

export type RegistrationState = "Registered" | "NotRegistered" | "Registering" | "Unregistering";
export type RegistrationPolicy = "RegistrationRequired" | "RegistrationOptional";

export type CloudShellProviderInfo = {
  id: string;
  namespace: string;
  authorizations?: ProviderAuthorization[];
  resourceTypes: ProviderResourceType[];
  registrationState: RegistrationState;
  registrationPolicy: RegistrationPolicy;
};
