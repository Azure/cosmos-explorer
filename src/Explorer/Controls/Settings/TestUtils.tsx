import ko from "knockout";
import * as DataModels from "../../../Contracts/DataModels";
import * as ViewModels from "../../../Contracts/ViewModels";
import Explorer from "../../Explorer";

export const container = new Explorer();

export const collection = {
  container: container,
  databaseId: "test",
  id: ko.observable<string>("test"),
  defaultTtl: ko.observable<number>(5),
  analyticalStorageTtl: ko.observable<number>(undefined),
  indexingPolicy: ko.observable<DataModels.IndexingPolicy>({
    automatic: true,
    indexingMode: "consistent",
    includedPaths: [],
    excludedPaths: [],
  }),
  rawDataModel: {
    uniqueKeyPolicy: {
      uniqueKeys: [
        {
          paths: ["/id"],
        },
      ],
    },
  },
  usageSizeInKB: ko.observable(100),
  offer: ko.observable<DataModels.Offer>({
    autoscaleMaxThroughput: undefined,
    manualThroughput: 10000,
    minimumThroughput: 6000,
    id: "offer",
    offerReplacePending: false,
  }),
  conflictResolutionPolicy: ko.observable<DataModels.ConflictResolutionPolicy>(
    {} as DataModels.ConflictResolutionPolicy,
  ),
  changeFeedPolicy: ko.observable<DataModels.ChangeFeedPolicy>({} as DataModels.ChangeFeedPolicy),
  geospatialConfig: ko.observable<DataModels.GeospatialConfig>({} as DataModels.GeospatialConfig),
  getDatabase: () => {
    return;
  },
  partitionKey: {
    paths: [],
    kind: "hash",
    version: 2,
  },
  partitionKeyProperties: ["partitionKey"],
  computedProperties: ko.observable<DataModels.ComputedProperties>([
    {
      name: "queryName",
      query: "query",
    },
  ]),
  vectorEmbeddingPolicy: ko.observable<DataModels.VectorEmbeddingPolicy>({} as DataModels.VectorEmbeddingPolicy),
  fullTextPolicy: ko.observable<DataModels.FullTextPolicy>({} as DataModels.FullTextPolicy),
  materializedViews: ko.observable<DataModels.MaterializedView[]>([
    { id: "view1", _rid: "rid1" },
    { id: "view2", _rid: "rid2" },
  ]),
  materializedViewDefinition: ko.observable<DataModels.MaterializedViewDefinition>({
    definition: "SELECT * FROM c WHERE c.id = 1",
    sourceCollectionId: "source1",
    sourceCollectionRid: "rid123",
  }),
  readSettings: () => {
    return;
  },
} as unknown as ViewModels.Collection;
