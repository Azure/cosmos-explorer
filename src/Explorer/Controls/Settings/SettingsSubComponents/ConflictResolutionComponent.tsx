import * as React from "react";
import { AccessibleElement } from "../../AccessibleElement/AccessibleElement";
import TriangleRight from "../../../../../images/Triangle-right.svg";
import TriangleDown from "../../../../../images/Triangle-down.svg";
import InfoBubble from "../../../../../images/info-bubble.svg";
import { StatefulValue } from "../../StatefulValue";
import * as ViewModels from "../../../../Contracts/ViewModels";
import * as DataModels from "../../../../Contracts/DataModels";
import Explorer from "../../../Explorer";

export interface ConflictResolutionComponentProps {
  collection: ViewModels.Collection;
  container: Explorer;
  tabId: string;
  conflictResolutionPolicyMode: StatefulValue<DataModels.ConflictResolutionMode>;
  conflictResolutionPolicyPath: StatefulValue<string>;
  conflictResolutionPolicyProcedure: StatefulValue<string>;
  onConflictResolutionChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onConflictResolutionLWWKeyPress: () => void;
  onConflictResolutionPolicyPathChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onConflictResolutionCustomKeyPress: () => void;
  onConflictResolutionPolicyProcedureChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

interface ConflictResolutionComponentState {
  conflictResolutionExpanded: boolean;
}

export class ConflictResolutionComponent extends React.Component<
  ConflictResolutionComponentProps,
  ConflictResolutionComponentState
> {
  public conflictResolutionPolicyModeCustomId: string;
  public conflictResolutionPolicyModeCRDTId: string;
  public conflictResolutionPolicyModeLWWId: string;

  constructor(props: ConflictResolutionComponentProps) {
    super(props);
    this.state = {
      conflictResolutionExpanded: true
    };
    this.conflictResolutionPolicyModeCustomId = `conflictResolutionPolicyModeCustom${this.props.tabId}`;
    this.conflictResolutionPolicyModeLWWId = `conflictResolutionPolicyModeLWW${this.props.tabId}`;
    this.conflictResolutionPolicyModeCRDTId = `conflictResolutionPolicyModeCRDT${this.props.tabId}`;
  }

  private toggleConflictResolution = (): void => {
    this.setState({ conflictResolutionExpanded: !this.state.conflictResolutionExpanded });
  };

  private getConflictResolutionModeComponent = (): JSX.Element => {
    return (
      <>
        <div className="formTitle">Mode</div>
        <div className="tabs" aria-label="Mode" role="radiogroup">
          <div className="tab">
            <AccessibleElement
              as="label"
              aria-label="ConflictResolutionLWWLabel"
              tabIndex={0}
              role="radio"
              aria-checked={
                this.props.conflictResolutionPolicyMode.current !== DataModels.ConflictResolutionMode.LastWriterWins
                  ? "true"
                  : "false"
              }
              className={`${this.props.conflictResolutionPolicyMode.isDirty() ? "dirty" : ""} ${
                this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.LastWriterWins
                  ? "selectedRadio"
                  : "unselectedRadio"
              }`}
              onActivated={this.props.onConflictResolutionLWWKeyPress}
            >
              Last Write Wins (default)
            </AccessibleElement>
            <input
              type="radio"
              name="conflictresolution"
              value={DataModels.ConflictResolutionMode.LastWriterWins}
              className="radio"
              id={this.conflictResolutionPolicyModeLWWId}
              onChange={this.props.onConflictResolutionChange}
              checked={
                this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.LastWriterWins
              }
            />
          </div>

          <div className="tab">
            <AccessibleElement
              as="label"
              aria-label="ConflictResolutionCutomLabel"
              tabIndex={0}
              role="radio"
              aria-checked={
                this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.Custom
                  ? "true"
                  : "false"
              }
              className={`${this.props.conflictResolutionPolicyMode.isDirty() ? "dirty" : ""} ${
                this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.Custom
                  ? "selectedRadio"
                  : "unselectedRadio"
              }`}
              onActivated={this.props.onConflictResolutionCustomKeyPress}
            >
              Merge Procedure (custom)
            </AccessibleElement>
            <input
              type="radio"
              name="conflictresolution"
              value={DataModels.ConflictResolutionMode.Custom}
              className="radio"
              id={this.conflictResolutionPolicyModeCustomId}
              onChange={this.props.onConflictResolutionChange}
              checked={this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.Custom}
            />
          </div>
        </div>
      </>
    );
  };

  private getConflictResolutionLWWComponent = (): JSX.Element => {
    return (
      <>
        <p className="formTitle">
          Conflict Resolver Property
          <span className="infoTooltip" role="tooltip" tabIndex={0}>
            <img className="infoImg" src={InfoBubble} alt="More information" />
            <span className="tooltiptext infoTooltipWidth">
              Gets or sets the name of a integer property in your documents which is used for the Last Write Wins (LWW)
              based conflict resolution scheme. By default, the system uses the system defined timestamp property, _ts
              to decide the winner for the conflicting versions of the document. Specify your own integer property if
              you want to override the default timestamp based conflict resolution.
            </span>
          </span>
        </p>
        <p>
          <input
            type="text"
            aria-label="Document path for conflict resolution"
            value={
              this.props.conflictResolutionPolicyPath.current === undefined
                ? ""
                : this.props.conflictResolutionPolicyPath.current
            }
            className={`${this.props.conflictResolutionPolicyPath.isDirty() ? "dirty" : ""}`}
            onChange={this.props.onConflictResolutionPolicyPathChange}
          />
        </p>
      </>
    );
  };

  private getConflictResolutionCustomComponent = (): JSX.Element => {
    return (
      <>
        <p className="formTitle">
          Stored procedure
          <span className="infoTooltip" role="tooltip" tabIndex={0}>
            <img className="infoImg" src={InfoBubble} alt="More information" />
            <span className="tooltiptext infoTooltipWidth">
              Gets or sets the name of a stored procedure (aka merge procedure) for resolving the conflicts. You can
              write application defined logic to determine the winner of the conflicting versions of a document. The
              stored procedure will get executed transactionally, exactly once, on the server side. If you do not
              provide a stored procedure, the conflicts will be populated in the
              <a
                className="linkDarkBackground"
                href="https://aka.ms/dataexplorerconflics"
                rel="noreferrer"
                target="_blank"
              >
                {` conflicts feed`}
              </a>
              . You can update/re-register the stored procedure at any time.
            </span>
          </span>
        </p>
        <p>
          <input
            type="text"
            aria-label="Stored procedure name for conflict resolution"
            value={
              this.props.conflictResolutionPolicyProcedure.current === undefined
                ? ""
                : this.props.conflictResolutionPolicyProcedure.current
            }
            className={`${this.props.conflictResolutionPolicyProcedure.isDirty() ? "dirty" : ""}`}
            onChange={this.props.onConflictResolutionPolicyProcedureChange}
          />
        </p>
      </>
    );
  };

  public render() {
    return (
      <>
        <AccessibleElement
          as="div"
          className="formTitle"
          onClick={this.toggleConflictResolution}
          onActivated={this.toggleConflictResolution}
          aria-expanded={this.state.conflictResolutionExpanded}
          role="button"
          tabIndex={0}
          aria-label="Conflict Resolution"
          aria-controls="conflictResolutionRegion"
        >
          {!this.state.conflictResolutionExpanded && (
            <span className="themed-images" id="ExpandChevronRightConflictResolution">
              <img
                className="imgiconwidth ssExpandCollapseIcon ssCollapseIcon"
                src={TriangleRight}
                alt="Show conflict resolution"
              />
            </span>
          )}

          {this.state.conflictResolutionExpanded && (
            <span className="themed-images" id="ExpandChevronDownConflictResolution">
              <img className="imgiconwidth ssExpandCollapseIcon" src={TriangleDown} alt="Show conflict resolution" />
            </span>
          )}
          <span className="scaleSettingTitle">Conflict resolution</span>
        </AccessibleElement>

        {this.state.conflictResolutionExpanded && (
          <div id="conflictResolutionRegion" className="ssTextAllignment">
            {this.getConflictResolutionModeComponent()}

            {this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.LastWriterWins &&
              this.getConflictResolutionLWWComponent()}

            {this.props.conflictResolutionPolicyMode.current === DataModels.ConflictResolutionMode.Custom &&
              this.getConflictResolutionCustomComponent()}
          </div>
        )}
      </>
    );
  }
}
