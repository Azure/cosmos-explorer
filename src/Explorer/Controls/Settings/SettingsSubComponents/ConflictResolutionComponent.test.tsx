import { shallow } from "enzyme";
import React from "react";
import { ConflictResolutionComponentProps, ConflictResolutionComponent } from "./ConflictResolutionComponent";
import { container, collection } from "../TestUtils";
import { StatefulValue } from "../../StatefulValue";
import * as DataModels from "../../../../Contracts/DataModels";

describe("ConflictResolutionComponent", () => {
  const props: ConflictResolutionComponentProps = {
    collection: collection,
    container: container,
    conflictResolutionPolicyMode: new StatefulValue<DataModels.ConflictResolutionMode>(
      DataModels.ConflictResolutionMode.Custom
    ),
    onConflictResolutionPolicyModeChange: () => {
      return;
    },
    conflictResolutionPolicyPath: new StatefulValue<"">(),
    onConflictResolutionPolicyPathChange: () => {
      return;
    },
    conflictResolutionPolicyProcedure: new StatefulValue<"">(),
    onConflictResolutionPolicyProcedureChange: () => {
      return;
    }
  };

  it("Sproc text field displayed", () => {
    const wrapper = shallow(<ConflictResolutionComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists("#conflictResolutionCustomTextField")).toEqual(true);
    expect(wrapper.exists("#conflictResolutionLwwTextField")).toEqual(false);
  });

  it("Path text field displayed", () => {
    props.conflictResolutionPolicyMode = new StatefulValue<DataModels.ConflictResolutionMode>(
      DataModels.ConflictResolutionMode.LastWriterWins
    );
    const wrapper = shallow(<ConflictResolutionComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists("#conflictResolutionCustomTextField")).toEqual(false);
    expect(wrapper.exists("#conflictResolutionLwwTextField")).toEqual(true);
  });
});
