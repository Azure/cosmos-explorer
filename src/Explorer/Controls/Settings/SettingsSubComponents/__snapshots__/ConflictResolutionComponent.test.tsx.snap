// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConflictResolutionComponent Path text field displayed 1`] = `
<Stack
  tokens={
    Object {
      "childrenGap": 30,
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Text>
      Mode
    </Text>
    <div
      aria-label="Mode"
      className="tabs"
      role="radiogroup"
    >
      <div
        className="tab"
      >
        <StyledLabelBase
          className="settingsV2Label  selectedRadio"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="radio"
          tabIndex={0}
        >
          Last Write Wins (default)
        </StyledLabelBase>
      </div>
      <div
        className="tab"
      >
        <StyledLabelBase
          className="settingsV2Label  unselectedRadio"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="radio"
          tabIndex={0}
        >
          Merge Procedure (custom)
        </StyledLabelBase>
      </div>
    </div>
  </Stack>
  <StyledTextFieldBase
    id="conflictResolutionLwwTextField"
    label="Conflict Resolver Property"
    onChange={[Function]}
    onRenderLabel={[Function]}
    styles={
      Object {
        "fieldGroup": Object {
          "borderColor": "",
          "height": 25,
          "selectors": Object {
            ":disabled": Object {
              "backgroundColor": "#ddd",
              "borderColor": "#969696",
            },
          },
          "width": 300,
        },
      }
    }
  />
</Stack>
`;

exports[`ConflictResolutionComponent Sproc text field displayed 1`] = `
<Stack
  tokens={
    Object {
      "childrenGap": 30,
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Text>
      Mode
    </Text>
    <div
      aria-label="Mode"
      className="tabs"
      role="radiogroup"
    >
      <div
        className="tab"
      >
        <StyledLabelBase
          className="settingsV2Label  unselectedRadio"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="radio"
          tabIndex={0}
        >
          Last Write Wins (default)
        </StyledLabelBase>
      </div>
      <div
        className="tab"
      >
        <StyledLabelBase
          className="settingsV2Label  selectedRadio"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="radio"
          tabIndex={0}
        >
          Merge Procedure (custom)
        </StyledLabelBase>
      </div>
    </div>
  </Stack>
  <StyledTextFieldBase
    id="conflictResolutionCustomTextField"
    label="Stored procedure"
    onChange={[Function]}
    onRenderLabel={[Function]}
    styles={
      Object {
        "fieldGroup": Object {
          "borderColor": "",
          "height": 25,
          "selectors": Object {
            ":disabled": Object {
              "backgroundColor": "#ddd",
              "borderColor": "#969696",
            },
          },
          "width": 300,
        },
      }
    }
  />
</Stack>
`;
